slapp/data_selection/select_data.py

```diff
@@ -0,0 +1,113 @@
+import slapp.utils.query_utils as qu
+import base64
+import os
+import argschema
+import marshmallow as mm
+import numpy as np
+
+
+insert_statement_template = (
+        "INSERT INTO experiment_selection "
+        "(base64_query_strings, slapp_commit_hash, ophys_experiment_ids, "
+        "sub_selected_ids, random_seed, comment_string) "
+        "VALUES ({}, {}, {}, {}, {}, {})")
+
+
+class DataSelectorSchema(argschema.ArgSchema):
+    query_strings = argschema.fields.List(
+        argschema.fields.Str,
+        required=True,
+        cli_as_single_argument=True,
+        description=("results will be the meta set of experiments. "
+                     "queries must return 'exp_id' key as alias for "
+                     "ophys_experiments.id"))
+    sub_selection_counts = argschema.fields.List(
+        argschema.fields.Int,
+        required=True,
+        cli_as_single_argument=True,
+        description="how many to sub-select from each query")
+    random_seed = argschema.fields.Int(
+        required=False,
+        default=42,
+        description="random number generator seed")
+    comment_string = argschema.fields.Str(
+        required=False,
+        default="",
+        description="logged in postgres entry as comment_string")
+
+    @mm.post_load
+    def check_lengths(self, data, **kwargs):
+        nstrings = len(data['query_strings'])
+        ncounts = len(data['sub_selection_counts'])
+        if nstrings != ncounts:
+            raise mm.ValidationError(f"{nstrings} query strings and "
+                                     f"{ncounts} sub_selection_counts "
+                                     "provided. These should be the same len")
+        return data
+
+
+class DataSelector(argschema.ArgSchemaParser):
+    default_schema = DataSelectorSchema
+
+    def run(self, lims_dbconn, label_dbconn):
+        self.logger.name = type(self).__name__
+
+        experiment_ids = []
+        b64_queries = []
+        for qstring in self.args['query_strings']:
+            experiment_ids.append(
+                    [i['exp_id'] for i in lims_dbconn.query(qstring)])
+            self.logger.info(
+                    f"{qstring}\n returned {len(experiment_ids[-1])} ids")
+            b64_queries.append(
+                base64.b64encode(bytes(qstring, 'utf-8')).decode('utf-8'))
+
+        rng = np.random.default_rng(self.args['random_seed'])
+
+        sub_experiments = []
```

Comment:
Why not [numpy.random.choice](https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html?highlight=random%20choice#numpy.random.choice)?

Reply:
You're right, instead of a shuffle and a selection, I can use choice with `replace=False`.
A slightly related point: numpy is moving away from `np.random.choice()` to 
```
rng = np.random.default_rng()
rng.choice()
```
https://numpy.org/doc/stable/reference/random/new-or-different.html#new-or-different

Reply:
switched to `choice` from `shuffle`